### ISSUE TITLE
Reintroducing auto-calibration

### ISSUE THREAD

**Issue by HazardCreative** [RotorHazard/RotorHazard#151]
Opening this issue as a central location to discuss how it could/should be done.

Now that we have data collection for each node, it seems like auto-calibration should be possible based on historical review. I'd like this calibration to consider previous races on the node as a starting point, but not rely on them. The algorithm should analyze the current race data as it comes in, and if it decides new calibration points are needed than it can choose to recalculate the portion of the race that has been recorded so far.

This method should allow the calibration to adapt to the different characteristics of quads and account for changes in node response from heat or other factors.

This adaptive algorithm should be the default, but optional.

I don't have any background in this kind of signal processing either, but with your help we've accomplished quite a lot so far. I'm confident we can do this.

**Comment by BoLtWD** [comment]
I like the option of auto calibration but feel it should be a option to turn on if and when needed so not to take away from manual controls.

**Comment by ScorpionTDL** [comment]
To be honest, I'm not a huge fan of a total automated approach.
I would prefer a guided callibration based on a historical data. 
For example adding the possibility to mark valid passes in the marshal page and the system automatically calculates the required values based on this information. 


Having it fully automated leaves a lot of guess work for the system which needs to be fine tuned with additional parameters. Without the information which peak was a valid pass its hard to make the correct guess 😉

**Comment by HazardCreative** [comment]
I don't intend to remove manual controls.

**Comment by Nandox7** [comment]
I like the idea to have an option that analyzes past data or current data during a race doing the system auto-calibration. But always allowing the user to later tweak them as needed.

Ideal the profiling would be done per pilot on a given event as often, even when all at running at a given power, antennas and VTx's differences make for different output levels.

**Comment by BoLtWD** [comment]
Yep that is what I like about the auto side. This would give any race director a chance to run a race out of the box allowing them to fin adjust later. Also like you say if you have a mixed bag of pilots all sharing channels during different rounds the auto side of things could help a director keep up with things.

**Comment by HazardCreative** [comment]
Yes, post-race adjustments will definitely stay.
Yes, I think the system should take into account a specific pilot's history if available. However, I'm not sure whether this should cross class boundaries or not. We run classes that expect the pilot to keep the same quad, and some where that's impossible.

**Comment by Nandox7** [comment]
That is a good point, even within the same class the same pilot can end up using more than 1 quad.

From a implementation point of view is just a matter of unlinking pilot details from the pilot registration id ornpilor seat (whatever you want to call it).

This is how many of other software work where you havr a global list of pilot details and you pull them in/out of a event&class "seat".

**Comment by BoLtWD** [comment]
Keep the same quad is not always the problem sometimes the VTX or Arial is beyond usable and they dont have a direct replacement to hand. In most cases I find that when you are running races when you have say 6 pilots racing and then the next 6 after all with different Quads you may find small differences that can effect the timer. With an auto tune it may help fine tune the differences out faster than a race director can manually adjusting .

**Comment by RogerBFPV** [comment]
There are 2 situations that result in the tuning needing to be altered. 
 1. As the nodes heat up etc, the RSSI values do sometimes shift a bit..    
 2. (more common) Having multiple quads utilizing the same frequency.   It is known that different VTX/antenna combinations can vary even when they are basically transmitting an average of 25mw.  

Using overall historical data for a specific node for tuning the node may help with 1. above,  but would not provide any value in helping solve 2. above.

Using historical data tied to a specific pilot 'may' help, although often the same pilot will fly a different quad, so 2. above is still problem.

It is known through my own experience as timing/scoring director at large events such as MultiGP sport class championship that the Immersion RC lapRF 8way can run for 12+ hours, with each channel seeing 100's of different quads in that time and NEVER needs to be re-calibrated.

While I do not have details of how Immersion has accomplished this, it does demonstrate that both 1. and 2. are solvable. Do they have better hardware or is it their software algorithms alone that allow for this?  I don't get the feeling they are using long historical data to accomplish this.

**Comment by HazardCreative** [comment]
I've never used an 8-way, but I definitely had tracking issues with the puck. I've heard it's the same tech inside, just one receiver instead of 8 of them. Inside the puck looks like a 5808 module, but it's connected to a patch antenna. The puck also doesn't have RF shielding. To date I've heard that adding antennas to a D5 is counter-productive, but I haven't explored this personally.

I've really pushed at my limits to get this as far as it has. I will certainly accept help in improving further. So far my efforts to reach out and find someone who knows this area better than myself have come up empty, but I haven't given up.
 
Question: How much effort is put into ensuring properly functioning VTx and antennas at your events? Each time I have personally experienced a quad falling under the tuning thresholds, I've brought out my power meter and found it to be significantly under-performing. But—like many RH users—I have a small group so my sample may not be representative.

**Comment by MoonKIE666** [comment]
The auto-calibration of the nodes can be a great help for race directors.

Since I started following this issue of chronometers based on RSSI readings, I have also heard many times that it is counterproductive to mount antennas in a system such as RotorHazard or Delta5. But the truth is that I have never seen any photograph of a DELTA5 or ROTORHAZARD that has mounted antennas, so I don't know where they get that argument.

I think that if you use good patch antennas like the ones mounted by IMMERSIONRC, with very little reception angle. That they were as addressable as possible, we could check how it works. I believe that the signal peaks when passing near the chrono would be much sharper and with that a history-based auto-calibration system could be better configured.

The problem is that we do not know which patch antenna IMMERSION uses, which I suppose will be its own. And on the subject of the algorithm, given the work you've been doing so far, I don't think it's the problem.

It would be great to have a telecommunications engineer among our ranks capable of designing a small patch antenna that could be used for this. Hahaha

**Comment by pulquero** [comment]
Crucially, it will need to be a LOW gain, highly directional antenna. The issue with a patch is that the gain might mean you pick up signals from neighbouring gates (unless perhaps you have the timer pointing skywards). It might also not be directional enough. Without the antenna, we achieve the low gain requirement, and shielding helps with the directionality.

Maybe Maarten Baert who openly licensed the triple feed patch can be persuaded to do an open licence dedicated timer antenna?

**Comment by HazardCreative** [comment]
The Immersion equipment uses a patch. 
![LapRF Puck](https://www.propwashed.com/wp-content/uploads/2018/02/MG_8496.jpg)
![LapRF 8-way](https://www.immersionrc.com/wp-content/uploads/2017/07/EventTimer600x400.jpg)

I don't think high-gain is an issue necessarily, but might require an additional attenuator. It's my understanding that a directional antenna could only be low-gain if it were extremely inefficient; such an antenna and would probably damage transmitter equipment so there's probably not a market for it. RF shielding may still necessary to prevent side-lobes.

**Comment by frizzle101101** [comment]
Ive been doing testing with a patch and there is too much gain overall, the best solution for directionality with the rx5808 modules as they are is to put q grounded box around the timer to reduce the gain it already has in all directions

**Comment by frizzle101101** [comment]
Imersiom rc uses a standard 5.8ghz patch which would give 8dbi of gain meaning that each of their nodes has very low attenuation or gain, they are somewhere reducing the overall gain of the system, but not by using an inefficent patch size
